File: 0235-lowest-common-ancestor-of-a-binary-search-tree/0235-lowest-common-ancestor-of-a-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        // Traverse the tree starting from root
        while (root != null) {
            // If both p and q are smaller, LCA lies in left subtree
            if (p.val < root.val && q.val < root.val) {
                root = root.left;
            }
            // If both p and q are greater, LCA lies in right subtree
            else if (p.val > root.val && q.val > root.val) {
                root = root.right;
            }
            // Otherwise, root is the split point → LCA
            else {
                return root;
            }
        }
        return null; // Should never happen if p and q exist in the tree
    }
}

File: 0235-lowest-common-ancestor-of-a-binary-search-tree/0235-lowest-common-ancestor-of-a-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
  * struct TreeNode {
   * int val;
    * TreeNode *left;
     * TreeNode *right;
      * TreeNode(int x) : val(x), left(NULL), right(NULL) {}
       * };
        */

        class Solution {
        public:
            TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
                    TreeNode* curr = root;
                            
                                    while (curr) {
                                                // If both nodes are greater than current, move to the right subtree
                                                            if (p->val > curr->val && q->val > curr->val) {
                                                                            curr = curr->right;
                                                                                        }
                                                                                                    // If both nodes are smaller than current, move to the left subtree
                                                                                                                else if (p->val < curr->val && q->val < curr->val) {
                                                                                                                                curr = curr->left;
                                                                                                                                            }
                                                                                                                                                        // If they split (one is left, one is right) or we hit p or q, this is the LCA
                                                                                                                                                                    else {
                                                                                                                                                                                    return curr;
                                                                                                                                                                                                }
                                                                                                                                                                                                        }
                                                                                                                                                                                                                return nullptr;
                                                                                                                                                                                                                    }
                                                                                                                                                                                                                    };